File: estruturas/pilha/Stack.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package mystack;

/**
 *
 * @author guest-BJ2kKf
 */
public class Stack implements IStack{
    private int size;
    private Object[] stack;
    private int constant;

    public Stack() throws Exception {
        this(0);
    }
    
    public Stack(int constant) throws Exception {
        if (constant < 0) {
            throw new EInvalidConstant();
        }
        this.constant = constant;
        this.size = 0;
        this.stack = new Object[20];
    }
     
    @Override
    public Object pop() throws Exception {
        if (this.isEmpty()) {
            throw new EEmptyStack();
        }
        Object o = this.stack[--size];
        this.stack[size] = null;
        return o;
    }

    @Override
    public void push(Object o) {
        if (size == this.stack.length) {
            if (this.constant == 0) {
                this.duplicate();
            }
            else {
                this.expand();
            }
        }
        this.stack[size++] = o;
    }

    @Override
    public Object top() throws Exception {
        if (this.isEmpty()) {
            throw new EEmptyStack();
        }
        return this.stack[size];
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public boolean isEmpty() {
        return this.size == 0;
    }
    
    private void expand() {
        this.stack = this.makeStack(this.stack, this.size + this.constant);
    }
    
    private void duplicate() {
        this.stack = this.makeStack(this.stack, this.size * 2);
    }
    
    private Object[] makeStack(Object[] stack, int newSize) {
        Object[] s = new Object[newSize];
        for (int i = 0; i < stack.length; i++) {
            s[i] = stack[i];
        }
        return s;
    }
}

File: estruturas/pilha-em-csharp/Stack.cs
using System;

namespace pilha_csharp
{
    class Stack : IStack
    {
        private int control;
        private Object[] stack;
        private int constant;

        private void updateStack(Object[] oldStack, int newLength) {
          this.stack = new Object[newLength];
          for (int i = 0; i < oldStack.Length; i++) {
            this.stack[i] = oldStack[i];
          }
        }

        private void handleFull() {
          if (this.control < this.stack.Length) {
            return;
          }
          int newLength;
          if (this.constant > 0) {
            newLength = this.stack.Length + this.constant;
          } else {
            newLength = this.stack.Length * 2;
          }
          this.updateStack(this.stack, newLength);
        }

        public Stack(int constant = 0) {
          this.control = 0;
          this.constant = constant;
          this.stack = new Object[20];
        }

        public void Push(Object o) {
          this.handleFull();
          int i = this.control;
          this.stack[i] = o;
          this.control++;
        }

        public Object Pop() {
          if (this.IsEmpty()) {
            throw new Exception("Stack is empty! Pop() not allowed.");
          }
          Object o = this.stack[this.control - 1];
          this.stack[this.control - 1] = null;
          this.control--;
          return o;
        }

        public Object Top() {
          if (this.IsEmpty()) {
            throw new Exception("Stack is empty! Top() not allowed.");
          }
          return this.stack[this.control - 1];
        }

        public int GetSize() {
          return this.control;
        }

        public bool IsEmpty() {
          return this.control == 0;
        }
    }
}
